Roundup of issues from recent PRs

guynir42 opened the pull request:
Rounding up a bunch of semi-related issues into a single PR, to cleanup a bunch of open issues we collected when working on a set of recent PRs. 

- Replace indexing of `upstream_images` for `new_image` and `ref_image` with a single ID of the ref image. Issue #178 
- Replace the `Image.header` with `Image.info` and `Image.raw_header` with `Image.header`. Issue #187 
- Remove limitations when making combined images, such as having different section ID or target. Issue #179 
- Allow images to not have a `section_id`. Also part of Issue #179 
- Add constraint on products that are unique with regards to `image_id` and `provenance_id`. Issue #160 
- Make sure products have their own `provenance_id` as part of their filenames (to make them unique). Issue #146 
- `downstream_images` as a lazy-loaded relationship on Image. Issue #151 
- Remove `recursive_merge()` in favor of regular merge and the new `merge_all()` that runs on Image, SourceList and Reference, because it needs to take special care of the products (which are not relationships!).

guynir42 commented:
In addition, and I don't know why this happened, I started to make some minor adjustments to how we merge things in the data store. That caused a whole bunch of bugs that I have been running after for three days, touching half of our code.

guynir42 commented:
@rknop There are still a couple of test failures, related to object cleanup (not the actual logic of the tests). 
If you have time to review this, you can start, and I will fix the remaining minor issues tomorrow morning.

rknop commented on models/cutouts.py:
```diff
@@ -44,6 +52,12 @@ def format(self, value):
         doc="The source list this cutout is associated with. "
     )
 
+    index_in_sources = sa.Column(
```
It occurs to me that there's a possible ambiguity here.  cutouts is associate with a source list, but the source lists on the new, ref, and sub will all have different indexes.  (And, it will often happen that there is no source in the ref's source list at this coordinate.)

I believe what's intended here is that this is the index in the sub's source list, and that that is the same thing that the sources_id points to.  We should make that clear in the documentation here.

(Assuming we're doing to do cutouts from warped ref, `and` that the ref was warped to the new, then the positions in all three images (ref, new, sum) will be the same.)

guynir42 replied:
Agreed. I will add some thing in the docstring to make it clear that the sources refer to the subtraction image. 
Maybe it would even be better to call it "detections"?

guynir42 replied:
I'm leaving it as `sources` because it is a `SourceList`, and using `detections` would make things much more confusing. I'll just add some emphasis in the docstrings.

rknop commented on models/image.py:
```diff
@@ -111,65 +111,46 @@ def format(self, value):
-        if self.new_image_index is None:
+        """Get the image that is NOT the reference image. This only works on subtractions (with ref+new upstreams)"""
+        image = [im for im in self.upstream_images if im.id != self.ref_image_id]
+        if len(image) == 0 or len(image) > 1:
```
If len(image) > 1, we should raise an exception here, rather than return None ; new_image() was called on a coadd rather than a subtraction, so new_image is meaningless.

guynir42 replied:
The problem is that `new_image` is not a function, it is a property. I think calling properties that routinely raise errors is going to be a headache (consider listing all properties of an  object). When it returns none you just know it is not relevant.

rknop commented on models/image.py:
```diff
@@ -487,8 +471,44 @@ def init_on_load(self):
+            new_image.sources = self.sources.merge_all(session=session)
+            new_image.wcs = new_image.sources.wcs
+            new_image.zp = new_image.sources.zp
+            new_image.cutouts = new_image.sources.cutouts
```
I think this subverts lazy loading.  

I think if self.cutouts is None, I don't think you want to set new_image.cutouts here.  (Same for other fields.)

guynir42 replied:
This is supposed to only set the attributes that were already there. 
You cannot lazy load cutouts or sources without specifying the provenance (which means it never gets lazy loaded, only explicitly loaded). 
In this case you have to do this for the merge to work. It essentially replaces the "recursive merge".

rknop commented on models/image.py:
```diff
@@ -788,8 +806,8 @@ def from_images(cls, images, index=0):
         output.end_mjd = max([image.end_mjd for image in images])  # exposure ends are not necessarily sorted
 
         # TODO: what about the header? should we combine them somehow?
+        output.info = images[index].info
```
We should probably do some filtering on info and header just like you did with the attributes.

Practically speaking, when we swarp the images together, it will strip all of the header information out; you can pass an argument to swarp telling it which header keywords to copy from the first image in the list.  We should update the image's header and info attributes after the swarp is done.  (Not sure if it does that right now, will have to look.)

guynir42 replied:
I have issue #172 to think about and sort this out.

rknop commented on models/image.py:
```diff
@@ -1071,7 +1099,7 @@ def invent_filepath(self):
                 section_id_int = int(self.section_id)
             except ValueError:
-                section_id_int = 0
+                section_id_int = 0  # TODO: maybe replace with a placeholder like 99?
```
If only there were an Int equivalent of NaN, that would be the right thing to use here.

When do we use the section_id_int?  Perhaps we should just embrace that section_id needs to be a string.

guynir42 replied:
The problem is that if you are using actual integer values (like CCD number) you'd want the (customizable) naming convention to allow you to do things like 01, 02... in the filename, and that only works with integers. 

The only problem is that same naming convention will be incorrect when `section_id=None`. I am not sure what to do about this, hence the TODO.

rknop commented on models/instrument.py:
```diff
@@ -1600,7 +1600,7 @@ def overscan_and_data_sections( self, header ):
         ----------
         header: fits.Header
-          The header of the image in question; use Image.raw_header not Image.header.
+          The header of the image in question; use Image.header not Image.info.
```
With the names .header and .info instead of .header and .raw_header, the comment after the semicolon probably isn't needed any more.  (No harm leaving it in.)

guynir42 replied:
yeah, I agree.

rknop commented on pipeline/catalog_tools.py:
```diff
@@ -384,7 +385,13 @@ def fetch_GaiaDR3_excerpt( image, minstars=50, maxmags=22, magrange=None, sessio
+                    if existing_catexp is None:
+                        session.add( catexp )  # add if it doesn't exist
+                    else:
+                        catexp = existing_catexp  # use the existing one
```
Hmm... I'm a little disturbed by this.  We're saving a catexp, but then returning an existing database entry.  It's entirely possible that the disk file will have changed, but we don't want to do that.

Looking further up in the code, I think that this should never happen; if there's already a db entry with the same filename, it should have been caught by the query above that searches for existing catalogs.

This should really be an exception; if the query on the corners, mag limits, and number of items didn't return anything, but there's an entry in the database that already has the filename of the catalog excerpt we made here, then there's a problem with our filename scheme not being sufficiently unique.

guynir42 replied:
I think I had some issues with this during tests (when the CatalogExcerpt survived the test restart but the files didn't, or the other way around). I'll try to replace it with an exception and see if the tests pass.

rknop commented on pipeline/data_store.py:
```diff
@@ -515,11 +515,11 @@ def get_image(self, provenance=None, session=None):
             # objects are wrong, but, oh well.)
             if (self.exposure_id is not None) and (self.section_id is not None):
-                if ( (self.image.exposure_id != self.exposure_id) or
+                if ( (self.image.exposure_id is not None and self.image.exposure_id != self.exposure_id) or
```
What's the reason for this change?  If self.image.exposure_id is None, then shouldn't self.exposure_id also be None?

guynir42 replied:
There are some situations (e.g., where Image is loaded from the test cache without the proper links) where it would fail before having a chance to assign all the new relationships.

rknop commented on pipeline/data_store.py:
```diff
@@ -1306,65 +1310,100 @@ def save_and_commit( self, exists_ok=False, overwrite=True, no_archive=False,
+            'detections',
+            'cutouts',
+            'measurements'
+        ]
```
This is another one of those cases where if we change the code somewhere else to add a new thing to the datastore, we have to remember to add an element to this list.

At least, let's put a comment in __init__ that warns of this.  It might also be good to make a class atribute _attributes_to_save or some such that holds this list, which sits at the top of the whole class, and which the comment in __init__ points to.

guynir42 replied:
I've refactored it into two lists at the beginning of the class definition. I still chose to explicitly initialize them in `__init__` because then my IDE doesn't complain that these attributes are defined outside the initialization.

rknop commented on tests/fixtures/decam.py:
```diff
@@ -209,7 +214,8 @@ def decam_raw_image( decam_exposure ):
     image.data = image.data[256:256+512, 256:256+512].copy()  # make it C-contiguous
-    return image
+
+    yield image
```
Is there any benefit to using `yield` rather than `return` when there's nothing in the function after the `yield`?  (I'd think if anything there's a benefit in using `return`, as it keeps it simpler; you don't have issues of calling it inside other things.)

guynir42 replied:
Some of the fixtures are nested, and if you have a return instead of a yield, it messes up the cleanup of the containing fixture.

rknop reviewed (changes requested):
Most of these cleanups look good.  I have a few questions and concerns, which are in the comments on the various lines of the code.

guynir42 commented:
@rknop I've addressed all comments. Please have another look and see if it is ok to merge (use the "see changes since last review" to save time). Thanks!
